spec/fixtures/generate_external_certs.sh

```diff
@@ -0,0 +1,322 @@
+#!/bin/bash
+
+# This script re-generates all of the keys and certs within the external_certs directory.
+# The files there are used by the spec/features/settings/external_cert_feature_spec test
+
+function init() {
+  ROOT_DIR=/tmp/ca_gen
+  CSR_DIR=$ROOT_DIR/csr
+  DIR_CERTS=$ROOT_DIR/certs
+  DIR_CONF=$ROOT_DIR/conf
+  DIR_KEYS=$ROOT_DIR/keys
+  DIR_CADATA=$ROOT_DIR/cadata
+  
+  mkdir -p $CSR_DIR
+  mkdir -p $DIR_CERTS
+  mkdir -p $DIR_CONF
+  mkdir -p $DIR_KEYS
+  mkdir -p $DIR_CADATA
+}
+
+function cleanup() {
+  rm -rf $ROOT_DIR
+}
+
+function init_ca() { # $1 ca_name
+  echo "Initiating CA  \"$1\""
+  generate_conf $1
+  # Create / clear certs directory for CA
+  mkdir -p $DIR_CADATA/$1/db.certs
+  rm $DIR_CADATA/$1/db.certs/* 2>/dev/null
+  
+  # Delete both db.index and db.index.attr if they exist
+  rm $DIR_CADATA/$1/db.inde* 2>/dev/null
+  
+  # Create blank db.index to start off CA empty
+  touch $DIR_CADATA/$1/db.index
+
+  # Create db.index.attr with default contents
+  echo unique_subject = yes > $DIR_CADATA/$1/db.index.attr
+
+  # Serial can start at 1 because serials are unique per CA
+  echo 01 > $DIR_CADATA/$1/db.serial
+  
+  # Create the keyfile for the CA
+  openssl genrsa -out $DIR_KEYS/ca_$1.key 2>/dev/null
+}
+
+function init_root_ca() { # $1 ca_name, $2 subject
+  init_ca $1
+  echo "Generating CRT \"$1\""
+  openssl req -new -x509 -days 100 -extensions v3_ca -key $DIR_KEYS/ca_$1.key -out $DIR_CERTS/ca_$1.crt -subj "$2"
+}
+
+function init_sub_ca() { # $1 ca_name, $2 parent_ca, $3 subject
+  init_ca $1
+  echo "Generating CRT \"$1\""
+  openssl req -sha256 -new -key $DIR_KEYS/ca_$1.key -out $CSR_DIR/ca_$1.csr -subj "$3"
+  openssl ca -batch \
+    -config $DIR_CONF/ca_$2.conf \
+    -extensions v3_ca \
+    -out $DIR_CERTS/ca_$1.crt \
+    -infiles $CSR_DIR/ca_$1.csr 2>/dev/null
+}
+
+function append_altnames() { # $1 altnames, $2 filename
+  # Split up dash delimited string into array
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "DNS.$(($index+1)) = ${altnames[index]}" >> $2
+  done
```

Comment:
It would be a little more efficient to put the concatenation after the while loop, so the file is only opened once instead of once per iteration.  Just move the `>> $2` after the `done`.

Also, in math expressions, variable names don't need the `$` prefix, which I think makes it a little easier to read.

Reply:
do/done blocks collect all of their stdout and can be redirected as a whole?

Reply:
Yup - from the `do` to the `done` is a block that can have redirection (including a pipe) applied after the `done`, kinda like its a separate script. Super handy for lots of things.

Reply:
Moved the append to after the done.